src/mbgl/style/style_parser.cpp

```diff
@@ -662,53 +664,151 @@ void StyleParser::parseLayers(JSVal value) {
             continue;
         }
 
-        // Parse paints already, as they can't be inherited anyway.
-        std::map<ClassID, ClassProperties> paints;
-        parsePaints(layerValue, paints);
-
-        util::ptr<StyleLayer> layer = std::make_shared<StyleLayer>(layerID, std::move(paints));
-
-        layers.emplace_back(layer);
-        layersMap.emplace(layerID, std::pair<JSVal, util::ptr<StyleLayer>> { layerValue, layer });
+        layersMap.emplace(layerID, std::pair<JSVal, util::ptr<StyleLayer>> { layerValue, nullptr });
+        ids.push_back(layerID);
     }
 
-    for (auto& pair : layersMap) {
-        parseLayer(pair.second);
-    }
-}
+    for (const auto& id : ids) {
+        auto it = layersMap.find(id);
 
-void StyleParser::parseLayer(std::pair<JSVal, util::ptr<StyleLayer>> &pair) {
-    JSVal value = pair.first;
-    util::ptr<StyleLayer> &layer = pair.second;
+        parseLayer(it->first,
+                   it->second.first,
+                   it->second.second);
 
-    if (value.HasMember("type")) {
-        JSVal type = value["type"];
-        if (!type.IsString()) {
-            Log::Warning(Event::ParseStyle, "layer type of '%s' must be a string", layer->id.c_str());
-        } else {
-            layer->type = StyleLayerTypeClass(std::string { type.GetString(), type.GetStringLength() });
+        if (it->second.second) {
+            layers.emplace_back(it->second.second);
         }
     }
+}
 
-    if (layer->bucket) {
+void StyleParser::parseLayer(const std::string& id, JSVal value, util::ptr<StyleLayer>& layer) {
```

Comment:
The `StyleParser` changes are necessitated by now needing to know the type of layer being instantiated, so that `StyleLayer::create` can create the appropriate subclass. So we have to delay that until later in the parsing process.